Create fips-compliance.md (#15870)
dotnet/docs: docs/standard/security/fips-compliance.md

---
title: .NET Core FIPS compliance
description: Explains .NET Core FIPS compliance
ms.date: 11/20/2019
---

# .NET Core FIPS compliance

.NET Core:

* Does **not** implement any cryptographic primitives.
* Passes cryptographic primitives calls through to the standard modules the underlying operating system provides.
* Does **not** enforce FIPS compliance in .NET Core apps.
* Does **not** stop developers from using non-FIPS compliant algorithms, even on systems configured for FIPS compliance.

It's the responsibility of the system administrator to configure FIPS compliance for an operating system.

If code is written for a FIPS-compliant environment, it's the responsibility of the developer to:

* Ensure that non-compliant FIPS algorithms are not used.

> * FIPS compliance is met.

Reviewer: Not in parallel grammatically to previous bullet.  Suggest:  "... the developer to ensure that: non-compliant FIPS algorithms are not used, FIPS compliance is met."

---
title: .NET Core FIPS compliance
description: Explains .NET Core FIPS compliance
ms.date: 11/20/2019
author: "Rick-Anderson"
ms.author: "riande"
---

# .NET Core FIPS compliance

> .NET Core:

Author: If @bartonjs  agrees, I could add the first paragraph of https://docs.microsoft.com/en-us/windows/security/threat-protection/fips-140-validation

> .NET Core:

Reviewer 2: > I could add the first paragraph of https://docs.microsoft.com/en-us/microsoft-365/compliance/offering-fips-140-2

Seems fine to me.

* Does **not** implement any cryptographic primitives.
* Passes cryptographic primitives calls through to the standard modules the underlying operating system provides.

> * Does **not** enforce FIPS compliance in .NET Core apps.

Reviewer 2: ```suggestion
* Does **not** enforce the use of FIPS Approved algorithms or key sizes in .NET Core apps.
```

> * Does **not** stop developers from using non-FIPS compliant algorithms, even on systems configured for FIPS compliance.

Reviewer 2: This seems redundant to the previous line, at least after my change.  Maybe I misunderstand one or the other.

> It's the responsibility of the system administrator to configure FIPS compliance for an operating system.

Reviewer 3: would this read better?

```suggestion
The system administrator is responsible for configuring the FIPS compliance for an operating system.
```

> If code is written for a FIPS-compliant environment, it's the responsibility of the developer to esure that:

Reviewer 3: ```suggestion
If code is written for a FIPS-compliant environment, the developer is responsible for ensuring that non-compliant FIPS algorithms aren't used.
```

> * Non-compliant FIPS algorithms are not used.

Reviewer 3: seems over the top to have a one bullet list

Author: Rewrite so it's not a long sentence. One bullet better than long sentence. Maybe @blowdart  or  @bartonjs  can suggest another bullet.

Reviewer 3: that's why I offered a rewrite to make it smaller.

Reviewer 2: "FIPS compliant" is a very difficult phrase.  In order to be "FIPS Compliant" you have to be using "FIPS Approved" algorithms from a "FIPS Certified" implementation (which frequently requires that implementation be running in a "FIPS Mode") -- but it only applies during a "FIPS Sensitive" (my phrase, not real) context.

e.g. FIPS, AFAIU, doesn't care if you use MD5 to set the Content-MD5 HTTP header (when the alternative is no data checksum).  But you can't do an RSA+MD5 signature, because MD5 can't be used in a cryptographic context.

Author: I changed the title to 
**Explains .NET Core Federal Information Processing Standard (FIPS) compliance**

FIPS folks won't have a problem. This article isn't aimed at new dev's looking to try something new.

For more information on FIPS compliance, see the following articles:

* [Windows FIPS Compliance](/windows/security/threat-protection/fips-140-validation)
* [Configuring Windows for FIPS Compliance](/windows/security/threat-protection/security-policy-settings/system-cryptography-use-fips-compliant-algorithms-for-encryption-hashing-and-signing)
* [OpenSSL FIPS Compliance](https://www.openssl.org/docs/fips.html)

> * [Configuring OpenSSL for FIPS Compliance](https://www.openssl.org/docs/fips/UserGuide.pdf)

Reviewer 2: I think there's a RHEL-specific one that we might want to link to.  The openssl.org documents are only for one variety of OpenSSL-FIPS.

Author: > I think there's a RHEL-specific one that we might want to link to. 
Can you provide a link?

Reviewer 4: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/security_guide/sect-security_guide-federal_standards_and_regulations-federal_information_processing_standard ?

Reviewer 2: Yep, that one.

> 

Reviewer 5: Fixes lint error.

```suggestion
```

---
title: .NET Core FIPS compliance
description: Explains .NET Core FIPS compliance
ms.date: 11/20/2019
author: "Rick-Anderson"
ms.author: "riande"
---

# .NET Core FIPS compliance

.NET Core:

> * Does **not** implement any cryptographic primitives.

Reviewer 2: This depends on what "any" and "primitives" mean.

We have an implementation of Password-Based Key Derivation Functions 1 and 2 (PBKDF1, PBKDF2).  Are those "primitives"?  Or Password-Based Encryption Schemes 1 and 2 (PBES1, PBES2)?

Author: Does **not** implement **MOST** ??

Author: With a few exceptions,  does **not** implement cryptographic primitives.

I could put a footnote and list what you provided if you 'd prefer a footnote.

Reviewer 2: Yeah, I'm not quite sure how we would phrase this here, but we can explicitly say that we use system implementations for AES, DES, ECDH, ECDSA, HMAC-MD5, HMAC-SHA1, HMAC-SHA256, HMAC-SHA384, HMAC-SHA512, MD5, RC2, RSA, SHA-1, SHA-256, SHA-384, SHA-512, or TripleDES (aka 3DES, aka DES-EDE).

We do have implementations of algorithms and schemes that use these cryptographic primitives, such as PBKDF1, PBKDF2, PBES1, PBES2, and HKDF.

Author: @bartonjs  we could delete the line
* Does **not** implement any cryptographic primitives.

* Passes cryptographic primitives calls through to the standard modules the underlying operating system provides.
* Does **not** enforce FIPS compliance in .NET Core apps.
* Does **not** stop developers from using non-FIPS compliant algorithms, even on systems configured for FIPS compliance.

It's the responsibility of the system administrator to configure FIPS compliance for an operating system.

If code is written for a FIPS-compliant environment, it's the responsibility of the developer to esure that:

* Non-compliant FIPS algorithms are not used.

> * FIPS compliance is met.

Reviewer 4: Not sure this line is needed, especially as the developer can't ensure that, it's a property of the OS configuration

Author: I added that to avoid the dreaded one-line bullet list. My D&I commitment is accessibility, and long sentences don't MT (machine translate). What's left is OK.

Reviewer 4: Ah that explains why my lovely formal writing has gone :)
